client/rpc/client.go

```diff
@@ -0,0 +1,164 @@
+
+	idsToMethods safesync.Map[serde.RequestMethod]
+}
+
+// NewRpcClient creates a new RpcClient
+func NewRpcClient(rpcServerUrl string, myAddress types.Address, chainId *big.Int, logger zerolog.Logger) *RpcClient {
+
+	nc, err := nats.Connect(rpcServerUrl)
+	handleError(err)
+	trp := natstrans.NewNatsTransport(nc, getTopics())
+
+	con, err := trp.PollConnection()
+	handleError(err)
+	nts := network.NewNetworkService(con)
+	nts.Logger = logger
+	c := &RpcClient{nts, myAddress, chainId, safesync.Map[chan interface{}]{}, safesync.Map[serde.RequestMethod]{}}
+	c.registerHandlers()
+	return c
+}
+
+// CreateLedger creates a new ledger channel
+func (rc *RpcClient) CreateLedger(counterparty types.Address, ChallengeDuration uint32, outcome outcome.Exit) directfund.ObjectiveResponse {
+
+	objReq := directfund.NewObjectiveRequest(
+		counterparty,
+		100,
+		outcome,
+		uint64(rand.Float64()), // TODO: Since numeric fields get converted to a float64 in transit we need to prevent overflow
+		common.Address{})
+
+	// Create a channel and store it in the responses map
+	// We will use this channel to wait for the response
+	resRec := make(chan interface{})
+	rc.responses.Store(string(objReq.Id(rc.myAddress, rc.chainId)), resRec)
+
+	requestId := rand.Uint64()
+	rc.idsToMethods.Store(fmt.Sprintf("%d", requestId), serde.DirectFundRequestMethod)
+
+	message := serde.NewJsonRpcRequest(requestId, serde.DirectFundRequestMethod, objReq)
+	data, err := json.Marshal(message)
+	if err != nil {
+		panic("Could not marshal direct fund request")
+	}
+	rc.nts.SendMessage(serde.DirectFundRequestMethod, data)
+
+	objRes := <-resRec
+	return objRes.(directfund.ObjectiveResponse)
+}
+
+func (rc *RpcClient) CloseLedger(id types.Destination) protocols.ObjectiveId {
+	objReq := directdefund.NewObjectiveRequest(id)
+
+	// Create a channel and store it in the responses map
+	// We will use this channel to wait for the response
+	resRec := make(chan interface{})
+	rc.responses.Store(string(objReq.Id(rc.myAddress, rc.chainId)), resRec)
+	requestId := rand.Uint64()
+	rc.idsToMethods.Store(fmt.Sprintf("%d", requestId), serde.DirectDefundRequestMethod)
+
+	message := serde.NewJsonRpcRequest(requestId, serde.DirectDefundRequestMethod, objReq)
+	data, err := json.Marshal(message)
+	if err != nil {
+		panic("Could not marshal direct fund request")
+	}
+	rc.nts.SendMessage(serde.DirectDefundRequestMethod, data)
+
+	objRes := <-resRec
+	return objRes.(protocols.ObjectiveId)
+}
+
+func (rc *RpcClient) Close() {
+	rc.nts.Close()
+}
+
+// registerHandlers registers error and response handles for the rpc client
+func (rc *RpcClient) registerHandlers() {
+
+	rc.nts.RegisterErrorHandler(func(id uint64, data []byte) {
+		panic(fmt.Sprintf("Objective failed: %v", data))
+	})
+
+	rc.nts.RegisterResponseHandler(func(id uint64, data []byte) {
+		rc.nts.Logger.Trace().Msgf("Rpc client received response: %+v", data)
+		method, reqFound := rc.idsToMethods.Load(fmt.Sprintf("%d", id))
```

Comment:
Would it be better to infer the method from the data itself, rather than keeping some state in the client to correlate requests ids to methods?

Reply:
That is a good alternative approach. Filed https://github.com/statechannels/go-nitro/issues/1098